File: golang/answer/answer_test.go
package answer

import (
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// expectedE2ECounts pins the shape of conformance/cases/e2e_hermetic.json. Both
// answer_test.go and askwith_test.go drive this fixture; neither may run against
// a silently shrunken corpus or case list.
var expectedE2ECounts = map[string]int{
	"corpus": 3,
	"cases":  4,
	"top_k":  5,
}

func TestE2EHermeticVectorCounts(t *testing.T) {
	var fixture struct {
		Corpus []Doc `json:"corpus"`
		TopK   int   `json:"top_k"`
		Cases  []struct {
			Question string `json:"question"`
		} `json:"cases"`
	}
	conform.Case(t, "e2e_hermetic.json", &fixture)
	for name, got := range map[string]int{
		"corpus": len(fixture.Corpus),
		"cases":  len(fixture.Cases),
		"top_k":  fixture.TopK,
	} {
		if want := expectedE2ECounts[name]; got != want {
			t.Errorf("e2e_hermetic.json %q: got %d, want %d", name, got, want)
		}
	}
}

// eqOptStr compares an actual value (present flag + value) against an expected
// nullable string from the fixture.
func eqOptStr(present bool, got string, want *string) bool {
	if want == nil {
		return !present
	}
	return present && got == *want
}

// TestAskConformance drives the full cite-or-abstain flow over the pinned §7
// hermetic e2e fixture: build the corpus once, then for EACH question assert
// Ask() yields the expected decision, answer, document, passage, and eu_id.
// Follows the tokenize exemplar: assert over ALL cases, no leniency.
func TestAskConformance(t *testing.T) {
	var fixture struct {
		Corpus []Doc `json:"corpus"`
		TopK   int   `json:"top_k"`
		Cases  []struct {
			Question string `json:"question"`
			Expected struct {
				Decision string  `json:"decision"`
				Answer   string  `json:"answer"`
				Document *string `json:"document"`
				Passage  *string `json:"passage"`
				EuID     *string `json:"eu_id"`
			} `json:"expected"`
		} `json:"cases"`
	}
	conform.Case(t, "e2e_hermetic.json", &fixture)

	if len(fixture.Cases) != expectedE2ECounts["cases"] {
		t.Fatalf("e2e_hermetic.json: got %d cases, want %d", len(fixture.Cases), expectedE2ECounts["cases"])
	}
	if len(fixture.Corpus) != expectedE2ECounts["corpus"] {
		t.Fatalf("e2e_hermetic.json: got %d corpus docs, want %d", len(fixture.Corpus), expectedE2ECounts["corpus"])
	}
	if fixture.TopK != expectedE2ECounts["top_k"] {
		t.Fatalf("e2e_hermetic.json: top_k = %d, want %d", fixture.TopK, expectedE2ECounts["top_k"])
	}

	for _, c := range fixture.Cases {
		res := Ask(fixture.Corpus, c.Question, fixture.TopK)

		if string(res.Evidence.Decision) != c.Expected.Decision {
			t.Errorf("%q: decision = %q, want %q", c.Question, res.Evidence.Decision, c.Expected.Decision)
		}
		if res.Answer != c.Expected.Answer {
			t.Errorf("%q: answer = %q, want %q", c.Question, res.Answer, c.Expected.Answer)
		}

		var (
			docPresent, passPresent, euPresent bool
			gotDoc, gotPass, gotEu             string
		)
		if len(res.Sources) > 0 {
			docPresent, gotDoc = true, res.Sources[0].Document
			passPresent, gotPass = true, res.Sources[0].Passage
		}
		if len(res.Claims) > 0 && len(res.Claims[0].Sources) > 0 {
			euPresent, gotEu = true, res.Claims[0].Sources[0]
		}

		if !eqOptStr(docPresent, gotDoc, c.Expected.Document) {
			t.Errorf("%q: document = %v (present=%v), want %v", c.Question, gotDoc, docPresent, c.Expected.Document)
		}
		if !eqOptStr(passPresent, gotPass, c.Expected.Passage) {
			t.Errorf("%q: passage = %v (present=%v), want %v", c.Question, gotPass, passPresent, c.Expected.Passage)
		}
		if !eqOptStr(euPresent, gotEu, c.Expected.EuID) {
			t.Errorf("%q: eu_id = %v (present=%v), want %v", c.Question, gotEu, euPresent, c.Expected.EuID)
		}
	}
}

File: js/src/answer/answer.test.ts
import { describe, it, expect } from "vitest";
import { loadCase } from "../conform/fixtures.js";
import { ask, type CorpusDoc } from "./answer.js";

// The ask flow is proven against the shared end-to-end fixture — every question
// must yield the expected decision, answer, and cited evidence. Follows the
// tokenize exemplar: load conformance/cases/e2e_hermetic.json, assert over ALL
// cases, no leniency.
interface Expected {
  decision: string;
  answer: string;
  document: string | null;
  passage: string | null;
  eu_id: string | null;
}

interface E2ECase {
  question: string;
  expected: Expected;
}

interface E2EFixture {
  corpus: CorpusDoc[];
  top_k: number;
  cases: E2ECase[];
}

/** Vector counts, pinned. A vector silently dropped from the fixture is a
 *  weakened contract that no per-case assertion can see. `top_k` is a scalar
 *  parameter, not a count, so it is pinned separately. */
const EXPECTED_COUNTS: Record<string, number> = { corpus: 6, cases: 8 };
const EXPECTED_TOP_K = 8;

describe("hermetic ask flow conformance", () => {
  const fixture = loadCase<E2EFixture>("e2e_hermetic.json");

  it("vector counts are pinned", () => {
    expect({ corpus: fixture.corpus.length, cases: fixture.cases.length }).toEqual(
      EXPECTED_COUNTS,
    );
    expect(fixture.top_k).toBe(EXPECTED_TOP_K);
  });

  for (const c of fixture.cases) {
    it(`ask(${JSON.stringify(c.question)})`, () => {
      const r = ask(fixture.corpus, c.question, fixture.top_k);
      const source = r.sources[0] ?? null;
      const euId = r.claims[0]?.sources[0] ?? null;

      expect(r.evidence.decision).toBe(c.expected.decision);
      expect(r.answer).toBe(c.expected.answer);
      expect(source?.document ?? null).toBe(c.expected.document);
      expect(source?.passage ?? null).toBe(c.expected.passage);
      expect(euId).toBe(c.expected.eu_id);
    });
  }
});
